group-coordinator/src/main/java/org/apache/kafka/coordinator/group/consumer/ConsumerGroup.java

```diff
@@ -0,0 +1,635 @@
+     *
+     * @return The epoch or -1.
+     */
+    public int currentPartitionEpoch(
+        Uuid topicId, int partitionId
+    ) {
+        Map<Integer, Integer> partitions = currentPartitionEpoch.get(topicId);
+        if (partitions == null) {
+            return -1;
+        } else {
+            return partitions.getOrDefault(partitionId, -1);
+        }
+    }
+
+    /**
+     * Compute the preferred (server side) assignor for the group while
+     * using the provided assignor for the member.
+     *
+     * @param oldMember The old member.
+     * @param newMember The new member.
+     *
+     * @return An Optional containing the preferred assignor.
+     */
+    public Optional<String> preferredServerAssignor(
+        ConsumerGroupMember oldMember,
+        ConsumerGroupMember newMember
+    ) {
+        // Copy the current count and update it.
+        Map<String, Integer> counts = new HashMap<>(this.serverAssignors);
+        maybeUpdateServerAssignors(counts, oldMember, newMember);
+
+        return counts.entrySet().stream()
+            .max(Map.Entry.comparingByValue())
+            .map(Map.Entry::getKey);
+    }
+
+    /**
+     * Returns the subscription metadata for all the topics whose
+     * members are subscribed to.
+     *
+     * @return An immutable Map containing the subscription metadata.
+     */
+    public Map<String, TopicMetadata> subscriptionMetadata() {
+        return Collections.unmodifiableMap(subscribedTopicMetadata);
+    }
+
+    /**
+     * Updates the subscription metadata. This replaces the previous one.
+     *
+     * @param subscriptionMetadata The new subscription metadata.
+     */
+    public void setSubscriptionMetadata(
+        Map<String, TopicMetadata> subscriptionMetadata
+    ) {
+        this.subscribedTopicMetadata.clear();
+        this.subscribedTopicMetadata.putAll(subscriptionMetadata);
+    }
+
+    /**
+     * Computes the subscription metadata based on the current subscription and
+     * an updated member.
+     *
+     * @param oldMember     The old member.
+     * @param newMember     The new member.
+     * @param topicsImage   The topic metadata.
+     *
+     * @return The new subscription metadata as an immutable Map.
+     */
+    public Map<String, TopicMetadata> computeSubscriptionMetadata(
+        ConsumerGroupMember oldMember,
+        ConsumerGroupMember newMember,
+        TopicsImage topicsImage
+    ) {
+        // Copy and update the current subscriptions.
+        Map<String, Integer> subscribedTopicNames = new HashMap<>(this.subscribedTopicNames);
+        maybeUpdateSubscribedTopicNames(subscribedTopicNames, oldMember, newMember);
+
+        // Create the topic metadata for each subscribed topic.
+        Map<String, TopicMetadata> newSubscriptionMetadata = new HashMap<>(subscribedTopicNames.size());
+        subscribedTopicNames.forEach((topicName, count) -> {
+            newSubscriptionMetadata.computeIfAbsent(topicName, __ -> {
```

Comment:
Is there a reason we use this method instead of put? Shouldn't each topic name be absent?

Reply:
It is a mistake. We can use `put` as you suggested.